Remove access to `CompletionEngine.completionToken`

datho7561 opened the pull request.

datho7561 committed:
Fix `ClassCastException` when working with module projects

- Remove `JavacFileObject`, use file objects from file manager

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Problem range for `IProblem.AbstractServiceImplementation`

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix some javadoc issue mapping

testforstephen committed:
Cache JCCompilationUnit from compilation job for reuse in problem converter (#853)

mickaelistria committed:
Set AccDeprected flag at DOM level

Fix https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/394

datho7561 committed:
Completion to add override for inherited method

eg. try completion at the `|` for the following cases:

```java
public Foo {
    |
}
```

```java
public Foo {
    toStr|
}
```

Closes #859

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
ASTConverter15JLS4Test.test0155 through test0160 - boxing and unboxing

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Add some problem mapping

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
More test annotations

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
index out of bounds exception

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
test0050/X.java - Cleaner handling of variable arity signatures

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Ensure parent packages are added to bindings

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Improvements to variable and method completion in method bodies

- Filter out private variables and methods from parent classes
- Filter out non-static variables and methods when completing inside a
  static method
- Distinguish between field completion and local variable completion
  (these have different icons in Eclipse)
- Set flags (eg. `public`, `protected`, `final`) for the completion items
  (these have different icons in Eclipse)

Fixes #861

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Make MatchLocator capable of working with DOM

Extra strategy in MatchLocator to use DOM, and tweaks in various
Locators to properly handle DOM. Tests in JavaSearchTests help a lot
(currently ~130/270 passing)

It basically clones and adapt some methods and adapt them to DOM,
A new flag is introduced to decide the strategy.

Done-ish (often with remaining issues)
* FieldLocator
* LocalVariableLocator
* MethodLocator
* SuperTypeReferenceLocator
* TypeDeclarationLocator
* TypeParameterLocator
* TypeReferenceLocator
TODO:
* AndLocator
* ModuleLocator
* PackageDeclarationLocator
* PackageReferenceLocator
* OrLocator
* VariableLocator

datho7561 committed:
Handle incomplete qualified names

eg. completion at the `|`

```java
public class HelloWorld {

    public static void main(String... args) {
        HelloWorld.|
    }
}
```

Handles static methods and members, as well as `this`, `super` and `class`.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion on `super.`

Fixes #865

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Map a problem id related to `super()`

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Method argument completion

Fill in method arguments with best guesses of the values,
using the existing mechanism.

- Refactor binding collection logic to filter out non-static members
  when applicable

Closes #868

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve annotation completion

- I broke Gayan's existing support in my previous patches (sorry Gayan), this PR fixes it
- Support completing right after `@`
- Improve annotation parameter completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Enable completion tests

- Fix some low hanging completion test errors

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Set debug info parameters (-g) to javac compiler

datho7561 committed:
Fix some completion relevances

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for keywords that can be inserted where statements can

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Add cancellation checks to DOMCompletionEngine

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Compile the generated sources into the expected target directory (#879)

datho7561 committed:
Use one default completion strategy

- Remove 'suitable binding' completion
- Fixes to package completionsome
  - Skip suggesting the default package
  - Use the same strategy whenever packages are suggested
  - Improve calculation of the prefix to make the matches more accurate
- Potentially address dupliate type completions

Fixes #870

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Share/deduplicate jar contents

Introduce a subclass of JavacFileManager that is capable of caching and
sharing jar contents.
Also ensure some references are removed from context upon usage to
encourage garbage collection.

Note that JDK content is still duplicated as it's using another
filemanager that seems harder to configure, using direct call to
`PlatformUtils.lookupPlatformDescription().getFileManager()`.
Also note that the cached entries only get disposed when all
filemanagers are closed/GCed.

datho7561 committed:
Begin reporting on completion progress monitor

See https://github.com/eclipse-jdtls/eclipse.jdt.ls/pull/3167#issuecomment-2407450906

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Mark the completion task as done using the progress monitor

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix test0080 - generics vs raw vs parameterized

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix test0164 - getDeclaringMethod() should return the declaration binding

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix NPE, convert 4 errors to fails

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Complete method return type, modifiers and thrown exceptions

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve ClassInstanceCreation completion

- recover `new ` in AST converter
- some fixes in bindings to get some test cases working
- scan backwards for first non-whitespace character when performing AST
  search
- use type hierachy to build a list of potential constructors
  - This method is expensive but accurate
- suggest creating anonymous classes when completing constructors for interfaces
- always provide all constructors of current type to be bug compatible
  with existing CompletionEngine
- use diamond operators

Limitations:
- relevance numbers are not quite right yet
- generic types don't quite work as expected

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Prevent error when renaming variable

Use a more advanced strategy when picking out which working copy to remove,
since sometimes the provided name of the file is only the last segment
of the path.

Fixes #884

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve field completion (`name.|`)

- Use bindings to build out a list of possible suggestions
- Handle access modifiers properly (`private`, `static`, `abstract`, ...)
  - Exclude members in grandparent types whose access was narrowed in
    parent types
- Fix completion preceeding another statement eg.

```java
{
  name.|
  Object myObject = null;
}
```
- Fix `super.to|`
- Prevent `super.|` and `super.to|` from suggesting methods that are
  abstract in the parent type

Fixes #891

Signed-off-by: David Thompson <davthomp@redhat.com>

snjeza committed:
JavacUtils.classpathEntriesToFiles doesn't configure dependent projects

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

datho7561 committed:
Fix some NPEs in completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Some fixes to keyword completion

- Perform prefix validation for keywords
- Suggest `super` when a statement is expected

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Avoid NumberFormatException when trying to get latest JRE.

akurtakov committed:
Really ignore compiler.warn.dangling.doc.comment

At some point the problem id that get ignored changed from 0 to -1 but
this warning code was missed in the conversion.
This saves a warning in every file with license header.

datho7561 committed:
Improve completion in extends/implements

- Change dom conversion logic to recover incomplete type names properly
- Restrict the suggested classes based on if they are a class or
  interface
- Do not suggest final classes
- Do not suggest the current class

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Prevent exception with multiline Javadoc method ref

Trim off the leading `*` when working on a method reference that spans
multiple Javadoc lines.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fixes to relevance and specific completion content

Makes changes to what fields in completion results we set
and what we set them to,
with the goal of reducing test failures

Signed-off-by: David Thompson <davthomp@redhat.com>

snjeza committed:
Improve JavacBindingResolver

akurtakov committed:
Map "compiler.warn.override.unchecked.ret"

IProblem.UnsafeReturnTypeOverride seems the correct one.

akurtakov committed:
Ignore 2 warnings without positions

These are compiler.note.removal.filename,
compiler.note.deprecated.plural.additional and the fact that they don't
have position info in them makes them not suitable to create Problems
for them.

datho7561 committed:
Fix some errors that occur when importing o.e.j.c incubator

- Do not reuse the context when compiling to multiple output directories
- Implement Alex's idea of a separate file of javac-specific error codes

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Remove false positive 'unused import' for import used in Javadoc

eg.

```java
import java.util.List;

/**
 * Like a {@link List} if it wasn't a list.
 */
public class NotAList {
}
```

The `java.util.List` import is needed to distiguish the Javadoc mention
of `List` from other potential `List` classes.

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Fix StringIndexOutOfBoundsException in Javadoc conversion

Make sure that calling substring is done with acceptable values, values
bigger than string length or negative ones are plain wrong.
```
!ENTRY org.eclipse.jdt.core 4 0 2024-10-31 11:45:52.859
!MESSAGE Failed to convert Javadoc
!STACK 0
java.lang.StringIndexOutOfBoundsException: Index 2912 out of bounds for
length 1736
```

datho7561 committed:
Field access and extends/implements completion improvements

- set prefix for `this.t|` completions correctly
- improve extends and implements completion with keywords
eg. avoid generic completion results and instead recommend keywords here:

```java
public class MyClass | {
	// ...
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
More fixes for field access

- Fix completion issue with `myVariable.|\nMyObj asdf = null;`
- Fix for `myTarget.|` completion

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Better support for ASTParser.setFocalPoint()

Reduce the parsed AST before resolving to get a performance boost.

akurtakov committed:
Fix wrong unused import for throws javadoc

akurtakov committed:
Do not generate false positive unused parameter warning

The following cases are fixed:
* Interface methods
* Abstract methods
* Methods marked with Override annotation

datho7561 committed:
Fix some regressions caused by field access fix

- `myMethod().|`
- completion after method declaration in `@interface`

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Share the FileManager used when -release is set

datho7561 committed:
Manage prefix for extends and implements completion

Recover the prefix in this case:

```java
public class MyClass ext| {
}
```

So that we know to only suggest `extends` and not `implements`.

Also, do not suggest `implements` for interfaces.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Error IDs for problems unique to javac

This causes 6 regerssions because 'possible this escape' is now
reported.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve completion for annotation members

- fix duplicates in annotation members
- fix range for annotation members
- fix type of annotation member completions
- opening completion in a normal annotation or single value annotation
  before `(`, eg `@MyAnnota|tion(value = 1)`,
  now tries to complete the annotation name rather than the annotation
  members

Fixes #930

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testTypeReference35 - corrupt jar on cp

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Some support for "proceedOnError"-like with javac

This allows to generate some .class files from invalid Java files,
enabling very convenient workflows during debug and hotswap.

datho7561 committed:
Improve completion for binary infix operator

In order to accomplish this,
I have changed the logic for adjusting the offset used for the node search.

This causes some regressions:
- A few cases where the test expects the AST not to be recovered but javac does recover it (eg. org.eclipse.jdt.core.tests.model.CompletionTests_1_5.test0197)
- @| in Javadoc is not handled. Before it got no completion due to not finding the correct node, but now it gets default completion, which causes some regressions. I'll fix these when dealing with @| properly
- The test cases sometimes expect the current class to be suggested in new | completion even when the type doesn't match, but sometimes they don't. I'll need to investigate further

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testTypeParameterConstructors01 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
No completion in strings, text blocks, comments, or Javadoc

- Javadoc will need to be revisited, but for now I turned off the
  default completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
[WIP] double colon completion

work-in-progress:
- prefix filtering
- see if I can get the binding key bugs resolved

Fixes #938

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix some NPEs

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Update doc

mickaelistria committed:
Restore completion strategy switch

datho7561 committed:
Handle resolving bindings for generic constructors

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for Javadoc tags

- Both block and inline tags
- Take into account which ones are applicable to the declaration being
  Javadocumented

Fixes #948

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Ignoring Javadoc of depedencies

When resolving, we do only need the symbols/bindings of depedencies.
Minimize the amount of option to skip eg parsing Javadoc for transitive
files that are only relevant as part of binding resolution.

mickaelistria committed:
Prevent from closing the Platform FileManager

as it's reused on the whole lifecyle of JDT.

mickaelistria committed:
Reduce processing using -implicit:none

* Also do not share the Platform filemanager with build (a bit more
isolation)
* a few other optimization to save CPU on irrelevant operations

mickaelistria committed:
Add output folders to classpath

Because resolving from .class deps is much faster than resolving from
.java deps

mickaelistria committed:
Skip linting without binding nor FORCE_PROBLEM_DETECTION

It's most likely useless to get linting when none of those are
requested.

Commit:
Fix testTypeDeclaration12 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix completion when typing out an import

eg.

```java
import java.lang.Li|

public class HelloWorld {
  public static void main(String... args) {
    System.out.println("Hello, world");
  }
}
```

Fixes #935

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testEnum05

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Ignore Javadoc for compiler

mickaelistria committed:
Do not lint when focusing on 1 particular position

mickaelistria committed:
Use partial DOM for completion, and for codeSelect on bigger files

mickaelistria committed:
Fix issue setting position of markdown comments

mickaelistria committed:
Fix unconfigure in p2.inf

datho7561 committed:
Fix replace range for anonymous class instance creation

Fixes #963

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Handle adding semicolon properly for import completion

eg. do not add it if it's already there, but add it if it isn't.

Also fixes a replace range bug for completion in this case:

```java
import java.util.Li|st;
```

Fixes #968

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testStaticImportMethod03 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Completion on unimported class

Fixes #964

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testStaticImportField04

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix NPE when converting constructor without body

Fixes #959

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve `throws` completion

- recover prefix properly
- don't suggest throws if the user has already added `throw` (i.e.
  misspelt it)
- don't suggest types when only `throws` is appropriate
- don't suggest interfaces and annotations when completing types
  in the `throws` section of a method declaration

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix issue populating multiple output class folders

mickaelistria committed:
Fix instruction.unconfigure

Commit:
Fix testTypeReference22

Signed-off-by: Rob Stryker <stryker@redhat.com>

Small change as per suggestion

Signed-off-by: Rob Stryker <stryker@redhat.com>

Oops - merge error

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
If outside the double quotes of a string literal, use the parent as the completion context

eg. complete `myString` in the following case:

```java
public class Main {

  public static void main(String... args) {
    String myString = "hi";
    System.out.println(|"hello");
  }

}
```

Fixes #966

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Avoid NPE when using completion for import QuickFix

Fixes #967

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve recovery on undefined method reference

datho7561 committed:
Fix error popups when class extends itself

Mostly null checks in bindings,
plus override type translation to prevent trying to access the broken
class.

Fixes #919

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix some CompletionContextTests

Commit:
Fix testTypeParameterConstructors04 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Support unknown expr in instanceof

datho7561 committed:
Suggest creating a new lambda when a funcitonal interface is needed

- many fixes to parameter completion
- implementation note:

given the following code snippet:

```
myMethod(12, , );
```

there are broken ast nodes for the 2nd and 3rd argument.
I use these nodes to help build completion when completing the 2nd and 3rd argument,
since if the broken nodes didn't exist it would be really difficult to figure out
which argument is being completed.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Make some match locator work with DOM

Commit:
Fix some conversion/bindings

Commit:
Endless loop during testBug229951a

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Keep string folding for build (not for CompilationUnitResolver)

datho7561 committed:
Expected type for ternary (`? :`)

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fixes to VariableDeclaration completion

- select the correct node to complete when working with variable
  declarations
- for FieldDeclarations, don't allow using declarations that occur after
  the current declaration when completing the initializer
  (eg. do not suggest `c` nor `b` in `int a = 1; int b = |; int c = 3;`)
  - also handle field declarations with multiple variable declaration
    fragments properly
    (eg. allow `a` as a completion in: `int a = 2, b = |`)

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Enable compilation progress reporting and cancellation

mickaelistria committed:
Revert "Fix error popups when class extends itself"

This reverts commit 56554987eb4806fb76e3d7ec1030ad8795364baf.

mickaelistria committed:
Single pass compilation even with multiple source folders

mickaelistria committed:
Fix possible IndexArrayOutOfBoundsError

datho7561 committed:
Completion for favourite static imports

Should get all the "favourite static imports" tests passing;
around 33 new passing tests and no regressions.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Do not close underlying channel for cached zip

By default, the underlying channels of a ZipFileSystem
gets closed if a consumer thread is aborted.
As we want to share the same object with multiple threads,
including some that might get aborted; we try to enforce
setUninterruptable() -with reflection until we find a better
way-.

Commit:
Historical merge error in javadoc, not new

Signed-off-by: Rob Stryker <stryker@redhat.com>

snjeza committed:
Add output folder to classpath (#1006)

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

mickaelistria committed:
Skip ProceedOnErrorTransTypes whenever possible

Keep track in the context of the files that have compile errors
so we only use ProceedOnErrorTransTypes on those and ignore it
for files without errors (as it can cause some other issues, eg
with XLargeTests being very slow)

Workaround for https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1008

mickaelistria committed:
Fix execution of Javac test

* Include all suites
* Fix testProceedOnErrors

mickaelistria committed:
More proceed on error: throw + var decl

mickaelistria committed:
Fix testBuildReferenceOtherProjectSource

test was calling Clean instead of Build

mickaelistria committed:
Fix some inner class not being copied

In some cases (such as switch expressions), Javac generates some extra
.class files of nested anonymous types. Those are only created duing the
"generate" phase, and are not detected during "analyze"; so we add in
the listener for Generate ability to record such newly generated
classes.

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1014

mickaelistria committed:
Include literal elements in literal

mickaelistria committed:
Bump parent version for org.eclipse.jdt.core.tests.javac

mickaelistria committed:
Prevent NPE in JavacBindingResolver

mickaelistria committed:
Avoid NPE in DOMToModelConverter.toLocalVariable

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1018

mickaelistria committed:
Map compiler.err.limit.string & compiler.err.limit.string.overflow pb

datho7561 committed:
Don't add import for `java.lang`

When importing a type via completion,
don't add an import for the type if it's from `java.lang`

A part of #1024

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Don't log possible syntax errors when editing

mickaelistria committed:
Convert compiler.err.switch.mixing.case.types

mickaelistria committed:
Some fixes to CompletionContextTests

mickaelistria committed:
Bump version of org.eclipse.jdt.core

datho7561 committed:
Don't adjust cursor offset when finding node to complete

Signed-off-by: David Thompson <davthomp@redhat.com>

snjeza committed:
javac doesn't configure output folders of dependent projects correctly

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

datho7561 committed:
Improve attached import completion for type completions

Only do it when the type isn't in the same class and use `null` instead
of a 0 length array

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testBug251827b and 5 others - import declarations java element must be found

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix some completionContextTests

mickaelistria committed:
JavacConverter: Do not set/fix position for recovered names

mickaelistria committed:
Revert "Don't adjust cursor offset when finding node to complete"

This reverts commit 170822b0f30bf19be087b7d97624875395595b24.
And it adapts it so that the if the resulting node after adjust is not a
SimpleName or a Literal (ie is structured), then this previous node is
used as context

mickaelistria committed:
Fix some contextCompletionTests (to 73)

Most issues related to token and positions are covered.

mickaelistria committed:
Some work on expected types (down to 49 failures)

mickaelistria committed:
Fix more CompletionContextTests (down to 43)

Commit:
Handle match files in the order they arrived

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Ensure all parent nodes fully encompass their child nodes

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Ensure calls to ASTParser.createBindings use the correct unit resolver

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
null check before passing a new TypeBinding to requestor

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Missing names must have length 0, not length 7 representing <error>

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Handle qualification when calculating relevance

I think I'm missing some edge cases, but this should fix a few cases.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Keep cached filesystems uniniterruptible

ZipFileSystem can have its underlying channel closed if a consuming
thread was interrupted while reading (eg aborted hover).
Override the fileManager so that the channel is not closed.

Commit:
Fix ASTConverter18Test.test0009 - illegal argument

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix NPE in testBug424138_001 and others; bad synchronized block

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Do not double-propose types

Sometimes, types were collected by the scrapeBindings method and the
type search.
This PR prevents suggesting the type from the type search if it was
already suggested through publishFromScope

Should fix ~10 test cases

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Translate problem id for unknown annotation member

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix position for ArrayInitializer

mickaelistria committed:
CompletionContextTest: null token after string literal

Commit:
NPE check in DOMCompletionEngine when erasure is null

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix regressions ASTConverterAST3Test.test0009 and others; also Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Used Resolved* JavaElement in Javac*Bindings.getJavaElement()

Those allow to store a more accurate key.

Commit:
Fixes #1038 - regression during some error syntax

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Camel case matching for completed types

- (when necessary)
- add .gitignore for tests Snjezana added in 08c4e999f3b277a70dba27db6f1261432bdcac77
- also, suggest types for the following case:

```java
public class HelloWorld {
    AA|
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Use resolved JavaElement from binding for CompletionContext.enclosing

+ Adjust bindings to fix errors highlighted by tests

datho7561 committed:
New method completion proposal

- Needs to do some string manipulation to avoid causing a regression for
  an edge case, unfortunately

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fixes CompletionContextTests.test0004 and possibly others

Signed-off-by: Rob Stryker <stryker@redhat.com>

rgrunber committed:
Remove calls to jobs that are no longer needed.

- jdt-ls-javac can be simulated by building JDT-LS with 'javac' profile
- Building JDT with 'javac' bundle in p2 repo is done by main job

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

snjeza committed:
Enable lombok with javac

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

mickaelistria committed:
Ensure bindings are ordered by depth

mickaelistria committed:
Use resolved key for binary elements too

+ suffix exception types

mickaelistria committed:
Strip exception from CompletionProposal.getSignature()

datho7561 committed:
Fixes to completing types

Focusing on the `CompletionTests.testType*` tests

- Fix relevance calculations: the R_CLASS / R_INTERFACE / R_ANNOTATION
  / etc. are only used when completing an `extends` or `implements`
  entry
- Complete keyword types in cases where types are suggested

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Constructor invocation completion proposal improvements

- Use expected types to infer constructor type
  (instead of existing logic which duplicates the logic of expected types)
- Use `IType` and `IMethod` when creating constructor invocation proposals,
  to avoid needing to create bindings
- Handle suggesting the default constructor when an `IType` has no
  explicit constructors
- Modify the content of the completion proposals to match what's
  expected by the tests

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix test0008 and others: empty block comment misunderstood by javac as block comment

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve method and type key

* Also include the exception from @throws in
JavacMethodBinding.getJavaElement() by reusing the existing matching
IMethod
* for resolved Wildcards that are unbound, make it same as `extends
java.lang.Object`)

mickaelistria committed:
Add boolean to ExpectedTypes when in conditions

mickaelistria committed:
add 1-click install link

mickaelistria committed:
Add feature for javac

mickaelistria committed:
Set NON_INHERITED on relevance

Start implementing conditions for
+ Some other fixes

datho7561 committed:
Fix expected type for return statements

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for `import module |;`

This is new Java 23 syntax.
This PR should make 6 new tests pass
(mostly from the java 23 completion suite)
This PR also incorporates
https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3479,
I forget specifically why, but it's needed.

- Fixes to the javac AST converter
  - I did a hack in the converter to recover the following case,
    since javac doesn't handle it currently:

```java
import module

public class HelloWorld {
}
```

- Adjust module completion relevance based on if it's required by the current
  module

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Make for fileObjects uninterruptible

In some cases, the system libraries still become unaccessible after an
interrupted thread.
This patch aims at ensuring the libraries are always made
uninterruptible upon access.

mickaelistria committed:
Do not suggest .class .super .this suffix for types

Those are only valid inside expresssions.

mickaelistria committed:
Force setAccessible in reflection

mickaelistria committed:
Open java.compiler/javax.tools=ALL-UNNAMED

as it's necessary in order to change visibility of field through
reflection.

mickaelistria committed:
Fix possible Javadoc parsing issue for reused nodes

The Javac AST may reuse nodes for different types inside Javadoc
MethodRef.
Fix the strategy to convert position so the the "official" position is
restored upon usage thus is read for reuse in other contexts.

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1075

mickaelistria committed:
Add javac feature to 2nd build step

Commit:
Cleanup and disable 3 tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix CompletionContextTests.test0116 and 1 other

Signed-off-by: Rob Stryker <stryker@redhat.com>

snjeza committed:
Add javax.tools to o.e.j.c.tests.javac

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

snjeza committed:
Fix ASTConverter15JLS4Test.test0008

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

datho7561 committed:
`@param |` completion

Complete type and method parameters after `@param` in Javadoc preceeding
type and method declarations

Signed-off-by: David Thompson <davthomp@redhat.com>

testforstephen committed:
Integrate javac library with a standalone bundle

mickaelistria committed:
Add support for packages in completion

mickaelistria committed:
Add completion of package keyword

datho7561 committed:
Fixes to Javadoc completion

- fix a bug related to ordering of tag names
- get `@param| ` completion to work as expected
- use `DOMCompletionUtil.findParent` instead of only checking the direct
  parent when calculating the node to use as context for completion
  (in the case of nodes within `Javadoc`)
- split out `MemberRef` and `MethodRef` to their own completion cases
  - for now these are stubs
- move the logic for type completion for `@see` and `@link` under the
  TagElement section, as it should be handled differently than the
  "default completion":
  - improve handling of qualified references
  - for the time being, suggest nothing when completing a type from a
    specific module.
    i.e. `@see java.base/java.util.|` no longer gives any results.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Do not create invalid JavacVariableBinding

when underlying VarSymbol is error

mickaelistria committed:
Filter proposed packages

mickaelistria committed:
Fix modifiers for types when coming from AST

mickaelistria committed:
Prevent NPE

mickaelistria committed:
Fix some JavacTypeBinding.getKey with generic extends

mickaelistria committed:
Resolve type for name when possible

mickaelistria committed:
Improve JavacMethodBinding.isRaw

mickaelistria committed:
Fix to ExpectedTypes

For Blocks and Lambdas

akurtakov committed:
Fix various warnings

* unused variables
* unqualified access
* switch from internal classes to api

mickaelistria committed:
Fix various possible NPEs

akurtakov committed:
Remove unused methods from JavacBindingResolver

mickaelistria committed:
Update README with updated args

mickaelistria committed:
Add NON_STATIC to relevance

mickaelistria committed:
Do not insert classname for static methods inside expression

datho7561 committed:
Implement completion for field ref in Javadoc

Note that there are two code paths: one through `MemberRef` and
another through an `@link` or `@see` `TagElement`.
This is because the AST conversion tests expect the following to be a
`TextElement` instead of a `MemberRef`:

```java
/**
 * @see MyClass#
 */
```

(and also:)

```java
/**
 * @see #
 */
```

However, the completion tests expect completion for MyClass' members to
work in this case.

Fixes #1103

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Add java.base/sun.nio.ch=ALL-UNNAMED to arguments

Also needed but missed.

mickaelistria committed:
Some work on type completion

* Type parameters
* Annotations
* Relevance

datho7561 committed:
Completion for method references in javadoc

- Fixes for the relevance numbers for methods in javadoc
- Suggest constructors (including the default constructor) in `@see` and
  `@link`
- Do not show completion when completing parameter name in method
  reference in Javadoc
- when referencing a method from Javadoc,
  and the cursor is in the method parameters,
  specifically right after a `(` or `,`,
  complete the entire method signature

This should fix 37 test cases

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improvements to completion

* Support type restriction for SwitchCase
* Support for QualifiedType
* Fix CompletionContext.getExpectedTypesSignatures()

akurtakov committed:
Record junit issues and warn on increase

mickaelistria committed:
Improve some import static type completion

mickaelistria committed:
Improve relevance computation for annotation without @Target

mickaelistria committed:
Work on completion with static import

* Return more proposals
* Do no repeat extra import in proposals

mickaelistria committed:
More support for annotation value names

datho7561 committed:
Extension point to replace `CompletionEngine`

Instead of using the
`-DCompilationUnit.codeComplete.DOM_BASED_OPERATIONS=true`,
make an extension point that allows replacing the CompletionEngine in `Openable`.
With the extension point,
we can contribute DOMCompletionEngine as a part of our javac fragment,
in a similar way to how it's being done for
`JavacCompilationUnitResolver`.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve support for completing enum constants in switch case

akurtakov committed:
Publish all issues in Jenkins

Commit:
Reverting removal of suspected typo, was not typo, was part of test

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Remove access to `CompletionEngine.completionToken`

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria commented:
What does it fix?
IMO, it's better to try using the CompletionEngine as utility class whenever possible rather than duplicating, as long as it's not using the Parser to proceed. Is it the case here?

datho7561 commented:
We cannot access CompletionEngine.completionToken if we are outside of the codeassist package, due to the current access level of the field. If we want to move the DOMCompletionEngine to the javac fragment, we will either need to change the access level or not access the field.

robstryker commented:
My thoughts: 

1) We shouldn't be aiming to move DOMCompletionEngine into javac fragment. We should be aiming (hopefully) to contribute it upstream. 

2) Using CompletionEngine rather than duplicating might make sense, but it also might not. If a full decoupling requires copying a very few simple and obvious methods with minimal logic, or if a copied version can be drastically minimized because we know the nested engine hasn't had any state added to it, then the simpler solution of copying and removing unused logic should be preferred.

Commit:
Remove dependency on CompletionEngine, excepting static methods

Signed-off-by: Rob Stryker <stryker@redhat.com>

akurtakov commented:
AFAICT the reason to move DomCompletionEngine to javac fragment is to make it possible to use our work with o.e.jdt.core bundle from I-build without modification. 
That is a must have for us to make it easier for people to start using it , it really doesn't matter whether it's javac fragment or some other ***our*** bundle as it's far from being upstreamable. Bundles coming https://github.com/eclipse-jdt should be perfectly fine for the project to work with.

datho7561 committed:
Collect `instanceof` and switch pattern variable bindings

Fixes #1130

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Fix unused import

As there are such cases often I raised it to error and enable the save
action.

akurtakov committed:
Obey SuppressWarnings("removal") in unused detector

Makes it slightly easier to use javac branch on big codebases as
hundreds of explicitly suppressed such warnings are now properly not
displayed anymore.

snjeza committed:
Unused imports are always listed as errors

datho7561 commented:
After adding @robstryker 's PR, I'm having second thoughts on the implementation he used. We need to subclass `InternalCompletionProposal`. ~~Why not override the methods that access the completion engine in the subclass? That way we don't need to make changes upstream.~~ Ah I see why, dumb idea, please ignore that

snjeza committed:
Test case for PR #1139

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

Commit:
NPE in binding resolver

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes 1145 - don't make a position when length is invalid

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Unnecessary import

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve converter & binding for IMemberValuePair

mickaelistria committed:
Set range for missing name anyway

properly set offset, with a length of 0.

mickaelistria committed:
Improve expectType for ArrayInitializer

datho7561 committed:
Completion for pattern variable initialized in for loop expression

eg. at the `|`:

```java
import java.util.List;
public class Main {
  private static record MyRecord(String a, int b) {}
  public static void main(String... args) {
    List myList = List.of(new MyRecord("hello", 12), new Object());
    for (int i = 0; i < myList.size() && myList.get(i) instanceof MyRecord(String jello, int kilometer); i++) {
      System.out.println(jel|);
    }
  }
}
```

Upstream doesn't handle this properly yet, so I added a test suite for
it.

Also handles collecing variables declared in the `for` initializer for
completion.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve some proceedOnError generation

mickaelistria committed:
Fix some relevance for member dereference

datho7561 commented:
@mickaelistria if you have some time, could you please take a look at this PR? This is a prerequisite for https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/pull/1129, since when we move DOMCompletionEngine to the javac fragment we won't be able to access CompletionEngine internals.

Commit:
Fixes #1155 - remove all ClassCastException in error log after importing org.eclipse.pde.ui

Signed-off-by: Rob Stryker <stryker@redhat.com>

Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

akurtakov committed:
Map compiler.warn.override.varargs.missing

akurtakov committed:
Ignore diagnostics without positions

"compiler.note.removal.plural.additional",
"compiler.note.unchecked.filename" don't have positions thus are not
usable in the ide.

akurtakov committed:
Fix failures of GHAs in local forks

As Javac needs Java 23 add it. While on it update Maven version used.

akurtakov committed:
Map one more varargs diagnostic

Cases like:
```
!ENTRY org.eclipse.jdt.core 4 0 2025-01-22 10:50:27.539
!MESSAGE Could not accurately convert diagnostic
(compiler.warn.varargs.unsafe.use.varargs.param)
/home/akurtakov/git/eclipse.platform.ui/tests/org.eclipse.jface.tests.databinding/src/org/eclipse/core/tests/databinding/observable/list/ListDiffTest.java:148:
warning: [varargs] Varargs method could cause heap pollution from
non-reifiable varargs parameter differences
		return Diffs.createListDiff(differences);
```
should be properly mapped now.

mickaelistria commented:
> We cannot access CompletionEngine.completionToken if we are outside of the codeassist package, due to the current access level of the field. If we want to move the DOMCompletionEngine to the javac fragment, we will either need to change the access level or not access the field.

Couldn't we just put the DOMCompletionEngine in the same package (even if moved into fragment)? That's already what we're doing for Bindings.

datho7561 commented:
> Couldn't we just put the DOMCompletionEngine in the same package (even if moved into fragment)? That's already what we're doing for Bindings.

You're right; we could do that. However, I still think it's important to decouple from CompletionEngine as much as possible. Many of the methods we are reusing from CompletionEngine are trivial to re-implement (e.g. they return a constant int). Other methods that we are currently reusing from CompletionEngine only work properly if the CompletionParser has been invoked and bunch of internal state has been populated, so we shouldn't be using them if we want correct results. If we rely too much on modifying the internal state of the nested CompletionEngine, the code becomes difficult to read and follow. For instance, I believe it's more clear what's happening when a static helper method calculates a list of completion results given some parameters than when we set a bunch fields on the nested completion engine then invoke a non-static method.

mickaelistria commented:
I think decoupling is eventually nice, but it's not a requirement and it's certainly more work. So IMO, at the moment, let's just use the same package. We may rework the coupling later when needed.

datho7561 commented:
If we decouple from CompletionEngine now, we prevent adding more coupling to the CompletionEngine in the short term. The more coupling we have, the harder it will be to decouple.

This PR fully decouples from the CompletionEngine; the work is already completed.

mickaelistria commented:
The proposed decoupling requires change in upstream APIs. The methods such as `incrementXXX`/`decrementXXX` are not conventional API and I don't think they'll be easily welcome in JDT.
While decoupling is nice, ability to work best with upstream with minimal change upstream is currently more critical.

akurtakov committed:
Add api filters

In order to have javac branch with bigger version of jdt.core so it
always overrides it during testing.

mickaelistria committed:
Forward release setting to AbstractImageBuilder

mickaelistria committed:
Test for release

akurtakov committed:
Add forgotten api filters

In order to have javac branch with bigger version of jdt.core so it
always overrides it during testing.

akurtakov committed:
Enable api-check and javadoc profiles

Should reduce problems in workspace.

Commit:
Dogfooding the pde project has revealed several NPEs

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Use JavaTask in JavaCompiler

Using JavaTask allows a more proper initialization, for example with
better support for release settings.

Commit:
Fixes #1170 - missing diagnostic constant

Signed-off-by: Rob Stryker <stryker@redhat.com>
